### PR TITLE
fix: remove aggressive timeout context setting middleware

### PR DESCRIPTION
**Pull request by zinic**
## Description

Remove a bunch of experimental machinery around controlling request context timeouts. This style of control leads to strange double-commit scenarios for the `http.Response` type and results in incorrect `504` responses where the client is behaving but behind a slow connection.

This changeset addresses https://github.com/SpecterOps/BloodHound/issues/13.

## Motivation and Context

There's a long history of debug attempts related to the code being cleaned up here. To save everyone a long read, the gist is that this was an experiment in aggressively trying to preserve UX for multi-user, single-tenant instances that back BloodHound Enterprise.

## How Has This Been Tested?

Integration test passes to validate that request handling operates as expected. One negative test harness written in Python that won't be committed.

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: cmd/api/src/api/middleware/middleware.go
Comment (superlinkx) [inline cmd/api/src/api/middleware/middleware.go:80]: Is there a reason to initialize requestID here? It seems like there's never a code path that will activate with it set to `ERROR` and if it did go off the rails, the zero value should be sufficient

**Review by superlinkx — COMMENTED** [review]
Just a quick question on a variable initialization change. Planning to pull this one down and test locally before approval.

File: cmd/api/src/api/middleware/middleware.go
Comment (zinic) [reply cmd/api/src/api/middleware/middleware.go:80]: Not really. I'll revert this. It was more just my preference.

Comment (zinic) [reply cmd/api/src/api/middleware/middleware.go]: Reverted :)

**Comment by juggernot325** [comment]
It looks like this middleware is the only place that used the `Prefer` header. Currently every endpoint in our OpenAPI documentation has this parameter specified. We should probably remove that to reflect this change in functionality. The question is, should we do it in this PR or a separate one?

File: cmd/api/src/api/middleware/middleware.go
Comment (superlinkx) [reply cmd/api/src/api/middleware/middleware.go]: Oh, I misread the error path. For some reason I read the `log.Errorf()` and assumed that was the end of the story. I still like this approach, but now it makes sense. I think either way it's written is fine now that I'm actually reading it properly.

**Comment by zinic** [comment]
The comment about the `Prefer` header is a good question. I have a sneaking suspicion that it still might be worth supporting. We could pull the `Prefer` header logic out in its own middleware so that it can set user deadline expectations in the request context for further consumption downstream without each handler having to parse or inspect the `Prefer` header.

I think a follow up discussion around use of the header and its documentation is worth having with the team. This bug is painful enough, however, that I'm not keen on blocking merge of this change for long.

**Comment by zinic** [comment]
After further internal team discussion, I'm going to rewrite portions of this PR to include better plumbing for the `Prefer` header so that users can bypass default transaction time-outs per-request.

**Comment by zinic** [comment]
This changeset has been updated with follow-up work:
* A new type in the BH context package has been added to track user timeout preferences.
* `Prefer` header is no longer tossed out but instead is now parsed and stored in the request `context.Context` value for future introspection. The context timeout is now correctly preserved and may be relied on for downstream context-aware function calls.
* Cypher queries now honor the `Prefer` header correctly.
* Early cancellation of responses has been removed.

File: cmd/api/src/queries/graph.go
Comment (juggernot325) [inline cmd/api/src/queries/graph.go:394]: Do we have any automated tests that verify this timeout value is being exercise correctly? I did a quick skim and did not see one, but I could have missed it. Whether we do or don't, do you think it would be valuable to add a test that also considers the `Prefer` header use case?

**Review by superlinkx — COMMENTED** [review]
Other than James' testing question, nothing really sticks out to me. Seems like a good way to replumb the Prefer header

File: cmd/api/src/queries/graph.go
Comment (zinic) [reply cmd/api/src/queries/graph.go:394]: Let me jam on this and update the changeset accordingly.

Comment (zinic) [reply cmd/api/src/queries/graph.go:394]: I've added three test cases that validate the inner QC timeout and user-preference overrides.

**Review by juggernot325 — APPROVED** [review]
LGTM! Thank you for adding those tests 💪

**Pull request merged**
